compiler/optimizer/UnsafeSubexpressionRemover.cpp

```diff
@@ -0,0 +1,207 @@
+      {
+      for (int32_t i = 0; i < node->getNumChildren(); ++i)
+         {
+         TR::Node *child = node->getChild(i);
+         bool didIt = anchorIfSafe(child, anchorPoint);
+         if (didIt && trace())
+            {
+            traceMsg(comp(), "  - Anchored child #%d %s n%d of %s n%d\n",
+               i,
+               child->getOpCode().getName(), child->getGlobalIndex(),
+               node->getOpCode().getName(), node->getGlobalIndex());
+            }
+         }
+      }
+   }
+
+bool
+OMR::UnsafeSubexpressionRemover::anchorIfSafe(TR::Node *node, TR::TreeTop *anchorPoint)
+   {
+   if (!isVisited(node))
+      anchorSafeChildrenOfUnsafeNodes(node, anchorPoint);
+
+   if (isUnsafe(node))
+      {
+      return false;
+      }
+   else
+      {
+      anchorPoint->insertBefore(TR::TreeTop::create(comp(), TR::Node::create(TR::treetop, 1, node)));
+      return true;
+      }
+   }
+
+void
+OMR::UnsafeSubexpressionRemover::eliminateStore(TR::TreeTop *treeTop, TR::Node *node)
+   {
+   // Time to delete the store.  This is more complex than it sounds.
+   //
+   // Generally, we want to replace the store with a treetop (ie. a
+   // no-op).  However, this is complicated by several factors, in
+   // order of increasing subtlety and complexity:
+   //
+   // 1. Stores can have multiple children -- as many as three for
+   //    an iwrtbar.  If the store is turning into a nop, we must
+   //    anchor all the other children.
+   // 2. Stores can appear under other nodes like CHKs and
+   //    compressedRefs.  Particularly for the CHKs, we must
+   //    perform the appropriate checks despite eliminating the
+   //    store, and so we must leave the store node in the
+   //    appropriate shape so its parents will do the right thing.
+   // 3. Some expressions are unsafe to evaluate after stores are
+   //    eliminated, and those expressions themselves must NOT be
+   //    anchored.
+   //
+   // In these transformations, we adopt the convention that we
+   // retain the original store whenever possible, modifying it
+   // in-place, and we retain its first child if possible.  These
+   // conventions are sometimes required, and so we just adopt them
+   // universally for the sake of consistency, efficiency, and
+   // readability in the jit logs.
+
+   // Anchor all children but the first
+   //
+   for (int32_t i = 1; i < node->getNumChildren(); i++)
+      {
+      TR::Node *child = node->getChild(i);
+      anchorIfSafe(child, treeTop);
+      child->recursivelyDecReferenceCount();
+      }
+
+   node->setNumChildren(1);
+
+   TR::Node *rootNode = treeTop->getNode();
+
+   //
+   // Note that the node at this point could be half-baked: an indirect
+   // store or write barrier with one child.  This is an intermediate
+   // state, and further surgery is required for correctness.
+
+   // Eliminate the store
+   //
+   if (rootNode->getOpCode().isSpineCheck() && treeTop->getNode()->getFirstChild() == node)
```

Comment:
I just noticed that I ought to have replaced `treeTop->getNode()` here with `rootNode`

Reply:
Corrected in commit 3863b194da815c3733e9e73243b0b7b147f030f9